data/summary/types/summary.go

```diff
@@ -0,0 +1,343 @@
+	d.HasFirstData = false
+	d.FirstData = nil
+
+	d.HasLastData = false
+	d.LastData = nil
+
+	d.HasOutdatedSince = false
+	d.OutdatedSince = nil
+}
+
+func (d *Dates) Update(status *UserLastUpdated, firstData time.Time) {
+	d.LastUpdatedDate = time.Now().UTC()
+
+	d.HasLastUploadDate = true
+	d.LastUploadDate = &status.LastUpload
+
+	d.HasFirstData = true
+	d.FirstData = &firstData
+
+	d.HasLastData = true
+	d.LastData = &status.LastData
+
+	d.HasOutdatedSince = false
+	d.OutdatedSince = nil
+}
+
+type Bucket[T BucketData, S BucketDataPt[T]] struct {
+	Date           time.Time `json:"date" bson:"date"`
+	LastRecordTime time.Time `json:"lastRecordTime" bson:"lastRecordTime"`
+
+	Data S `json:"data" bson:"data"`
+}
+
+type BucketDataPt[T BucketData] interface {
+	*T
+	CalculateStats(interface{}, *time.Time) (bool, error)
+}
+
+func CreateBucket[T BucketData, A BucketDataPt[T]](t time.Time) *Bucket[T, A] {
+	bucket := new(Bucket[T, A])
+	bucket.Date = t
+	bucket.Data = new(T)
+	return bucket
+}
+
+type Buckets[T BucketData, S BucketDataPt[T]] []Bucket[T, S]
+
+type Stats interface {
+	CGMStats | BGMStats
+}
+
+type StatsPt[T Stats] interface {
+	*T
+	GetType() string
+	GetDeviceDataType() string
+	Init()
+	GetBucketsLen() int
+	GetBucketDate(int) time.Time
+	Update(any) error
+}
+
+type Summary[T Stats, A StatsPt[T]] struct {
+	ID     primitive.ObjectID `json:"-" bson:"_id,omitempty"`
+	Type   string             `json:"type" bson:"type"`
+	UserID string             `json:"userId" bson:"userId"`
+
+	Config Config `json:"config" bson:"config"`
+
+	Dates Dates `json:"dates" bson:"dates"`
+	Stats A     `json:"stats" bson:"stats"`
+}
+
+func NewConfig() Config {
+	return Config{
+		SchemaVersion:            1,
+		HighGlucoseThreshold:     highBloodGlucose,
+		VeryHighGlucoseThreshold: veryHighBloodGlucose,
+		LowGlucoseThreshold:      lowBloodGlucose,
+		VeryLowGlucoseThreshold:  veryLowBloodGlucose,
+	}
+}
+
+func (s *Summary[T, A]) SetOutdated() {
+	s.Dates.OutdatedSince = pointer.FromTime(time.Now().UTC())
+}
+
+func NewDates() Dates {
+	return Dates{
+		LastUpdatedDate: time.Time{},
+
+		HasLastUploadDate: false,
+		LastUploadDate:    nil,
+
+		HasFirstData: false,
+		FirstData:    nil,
+
+		HasLastData: false,
+		LastData:    nil,
+
+		HasOutdatedSince: false,
+		OutdatedSince:    nil,
+	}
+}
+
+func Create[T Stats, A StatsPt[T]](userId string) *Summary[T, A] {
+	s := new(Summary[T, A])
+	s.UserID = userId
+	s.Stats = new(T)
+	s.Stats.Init()
+	s.Type = s.Stats.GetType()
+	s.Config = NewConfig()
+	s.Dates = NewDates()
+
+	return s
+}
+
+func GetTypeString[T Stats, A StatsPt[T]]() string {
+	s := new(Summary[T, A])
+	return s.Stats.GetType()
+}
+
+func GetDeviceDataTypeString[T Stats, A StatsPt[T]]() string {
+	s := new(Summary[T, A])
+	return s.Stats.GetDeviceDataType()
+}
+
+type Period interface {
+	BGMPeriod | CGMPeriod
+}
+
+func AddBin[T BucketData, A BucketDataPt[T], S Buckets[T, A]](buckets *S, newStat Bucket[T, A]) error {
+	var hourCount int
```

Comment:
Not pythony, no. Most of the time I do this for clarity of scope, because `if` etc are scoped, its clearer for me to define them at the top so that I can be sure im never creating scoped versions later by accident.

in this case, i can clean this up without too much worries, as we arent nesting far, though i will use `var` instead of `:=` to prevent any accidental re-declares.